utf8: don't linearize cells for validation

avikivity opened the pull request:
Currently, we linearize large UTF8 cells in order to validate them. This can cause large latency spikes if the cell is large.

This series changes UTF8 validation to work on fragmented buffers. This is somewhat tricky since the validation routines are optimized for single-instruction-multiple-data (SIMD) architectures.

The unit tests are expanded to cover the new functionality.

Fixes #7448.

haaawk commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
 
 namespace utf8 {
 
```
Can't comment on commit message so posting here:
s/that **is** spans multiple buffers/that spans multiple buffers

avikivity replied:
fixed

haaawk commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
+inline
+evaluate_codepoint(const uint8_t* data, size_t len) {
+    auto byte1 = data[0];
+    static uint8_t len_from_first_nibble[16] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
```
std::array? and const?

avikivity replied:
std::array gives nothing here IMO. It's useful to wrap an array in a regular object without special behavior when passing it around, especially to templates. Here it's just obfuscation.

Added const.

avikivity replied:
> I think the validator was missing some important checks and didn't understand other checks it did - see my review.

This series does not introduce a validator, so comments about the validator do not apply.

> If any of the potential bugs I raised are true, it also means the tests missed them so they aren't exhaustive enough.
> Moreover, I think the fragmented utf8-checking should actually be simpler than the way you did it. If a UTF-8 "code point" is split between buffers, there is no reason to merge the two halfs. The first half includes the first byte - which indicates the code's **length**, and all the rest of the bytes simply need to begin by bits 10 - i.e., be between 0x80 and 0xBF. 

That's not true. 

Edit: actually it is true. So maybe it can be simplified.

>In other words, in the second buffer, we just need to check that the first M (calculated in the previous buffer) bytes are between 0x80 and 0xBF - there is no need to keep any byte data from the previous buffer.

avikivity replied:
Edit2: it's not true.The sequence f0 80 80 80 codes for codepoint 0, but is illegal.

haaawk replied:
I was under the impression that in modern C++ std::array should always replace old arrays when the size is known. Isn't std::array safer?

avikivity replied:
No, it just wraps a regular C array.

It's safer when you pass it around, since C arrays have strange rules where they decay into a pointer and lose information about the size.

haaawk replied:
Fair enough. It's not safer. It still seems more idiomatic in modern c++. Because why would anyone use C arrays when there is std::array and std::vector?

nyh replied:
> Edit2: it's not true.The sequence f0 80 80 80 codes for codepoint 0, but is illegal.

I didn't think about this. But not all is lost. Basically, leading zeros are not allowed. So it's not enough to remember remaining_bytes for the current codepoint, we also need to remember is_second_byte which means whether the next byte to be read is the first one between 0x80 and 0xBF - and if it is the first one, it must not be 0x80.

Anyway, I see now that https://tools.ietf.org/html/rfc3629 has in "Syntax of UTF-8 Byte Sequences" the same weird ranges that I don't understand. Maybe I'm missing something important, but I don't understand yet what :-(

avikivity replied:
Yes, but I don't want to invent a new mechanism for validating the buffer-straddling codepoint. There's no real benefit, and there is a risk of getting the validation wrong.

haaawk commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
+codepoint_status
+inline
+evaluate_codepoint(const uint8_t* data, size_t len) {
+    auto byte1 = data[0];
```
const?

avikivity replied:
Why? const it useful on interfaces, it tells the reader of the interface that the callee does not change the value.

How does it help here?

haaawk replied:
when a variable is const a reader knows it will never change. When it's not const a reader has to read the whole code to know.
If something is const like byte1 in our case why not mark it as such?

avikivity replied:
Ok, I'll add const. I think it's noise in small functions, but if it helps someone, it's worth it.

haaawk commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
-        } else if (len >= 2 && byte1 >= 0xC2 && byte1 <= 0xDF &&
+            return codepoint_status{.bytes_validated = codepoint_len};
+        } else if (byte1 >= 0xC2 && byte1 <= 0xDF &&
                 (int8_t)data[1] <= (int8_t)0xBF) {
```
Shouldn't we also check (int8_t)data[1] >= 0x80?

avikivity replied:
There is no (int8_t)data[1] that can be < (int8_t)0x80 - that's the minimum int8_t value. A nice trick by the original author.

In any case it is not related to this refactoting.

haaawk replied:
Nice trick indeed.

haaawk commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
+        } else {
             const uint8_t byte2 = data[1];
 
             // Is byte2, byte3 between 0x80 ~ 0xBF
```
This comment says that we check bytes are within [0x80, 0xBF] but the check below compares only against 0xBF. Shouldn't we also compare against 0x80?

avikivity replied:
All int8_t values are >= 0x80, that is the minimum value.

haaawk replied:
It's not that all int8_t values are >= 0x80. It's more that every value that's < 0x80 in uint8_t is bigger than 0xBF in int8_t. Otherwise, the check wouldn't work.

But you're right everything's fine.

avikivity replied:
godbolt thinks it is:

   https://godbolt.org/z/9rKah7

0xbf checks the other side of the interval. The trick reduced a closed interval into a half-open interval.

haaawk replied:
Compiler optimization is understandable. 0x80 is the minimal value of int8_t so there's no point checking that something is smaller than 0x80. But it's all about the conversion from uint8_t to int8_t. Values that are < 0x80 in uint8_t after conversion to int8_t become > than 0xBF and that's why this trick works. If values that are < 0x80 in uint8_t after conversion to int8_t were taking values smaller than 0xBF then the trick wouldn't work. The `<= 0xBF` check serves two purposes:
1. eliminate all values that in uint8_t are bigger than 0xBF because they are still bigger than 0xBF after conversion to int8_t
2. eliminate all values that in uint8_t are smaller than 0x80 because after conversion to int8_t they become bigger than 0xBF

What I'm saying is that the fact that 0x80 is a minimal value for int8_t does not make the trick correct. What does is the way uint8_t is converted to int8_t. It's based on the fact that the conversion turns every x < 0x80 into y > 0xBF. If it wasn't true the trick wouldn't work even with 0x80 being the minimal value of int8_t.

nyh replied:
```
const int byte2_ok = (int8_t)byte2 <= (int8_t)0xBF;
```
Wow, what a trick... Basically, 0xBF is a negative number, so for byte2 to be smaller than it, it needs to be negative too (i.e., the unsigned version is >= 0x80), and must also be <= 0xBF. This should have been in a separate function with clear comments explaining how/why it works. I agree it's pre-existing code, so you don't *have* to fix it too, but maybe it's a good opportunity since we understand it now...

haaawk replied:
I agree it's very hacky. Hard to spot.

avikivity replied:
The same trick is used in the SIMD code, where it's not possible to extract.

haaawk commented on utils/utf8.cc:
```diff
@@ -87,8 +95,9 @@ static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t l
+                return codepoint_status{.bytes_validated = codepoint_len};
+            }
+            {
                 // Is byte4 between 0x80 ~ 0xBF
```
Again, compares only against 0xBF. 0x80 is never checked.

avikivity replied:
Again, the codeeis tricky but correct.

haaawk replied:
True.

haaawk commented on utils/utf8.cc:
```diff
@@ -87,8 +95,9 @@ static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t l
                      (byte1 >= 0xEE && byte1 <= 0xEF))) {
-                bytes = 3;
-            } else if (len >= 4) {
+                return codepoint_status{.bytes_validated = codepoint_len};
```
I have to admit, I don't understand those checks. For example `(byte1 == 0xE0 && byte2 >= 0xA0)` why is `byte2 >= 0xA0` needed? I would understand `byte2 >= 0x80` but have no idea what `0xA0` is.
Same with the comparison of byte1 against `0xE1`, `0xED`, and `0xEE`. I thought that meaningful values of byte1 are `0x00 - 0x7F`, `0xC0-0xDF`, `0xE0-0xEF`, and `0xF0-0xF7`?

nyh replied:
I don't understand this either. For 3 bytes, byte0 should have binary 1110..., in other words between 0xE0 and 0xEF. What's all these special cases for 0xED and 0xEE?
All this code seems to me like maybe we're overthinking things... IIRC, in UTF-8, the first byte determines the length (that's the codepoint_len thing), but then - the rest of the bytes just need to begin with bits 10, i.e., between 0x80 and 0xBF.  That's it. Why do we have any other case?

avikivity replied:
First, these comments don't concern the refactoring. I did not change the logic, just the interfaces.

Second, the code is very careful and correct. While the first byte determines the length, the following bytes have to follow a distinct pattern, and if they don't, the sequence is invalid. They don't just have to fit the range 0x80-0xbf.

I recommend you study the code, not the changes, if you want to understand the validation logic. My series does not change validation in any way (at least not intentionally), it just makes it work for scattered buffers.

haaawk commented on utils/utf8.hh:
```diff
@@ -64,6 +65,54 @@ inline std::optional<size_t> validate_with_error_position(bytes_view string) {
 }
 
+std::optional<size_t> validate_with_error_position_fragmented(const FragmentRange auto& fr) {
+    uint8_t partial_codepoint[4];
```
std::array?

avikivity replied:
why?

haaawk replied:
better safety and more idiomatic for modern c++

haaawk commented on utils/utf8.hh:
```diff
@@ -64,6 +65,54 @@ inline std::optional<size_t> validate_with_error_position(bytes_view string) {
+                }
+                bytes_validated += partial_filled;
+                partial_filled = partial_more_needed = 0;
+            }
+            if (!len) {
```
Super nit: it is guaranteed that after a while loop above either partition_more_needed == 0 or len == 0 so here you can just do else and inside this else do assert (!len).

haaawk replied:
Ok actually no. It is also possible for both partial_more_needed and len to be zero so `else` wouldn't work.

avikivity replied:
Correct.

haaawk reviewed (approved):
LGTM I just have a couple of questions about the first commit. Not really related to this change.

nyh commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
+    static uint8_t len_from_first_nibble[16] = { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 3, 4 };
+    auto codepoint_len = len_from_first_nibble[byte1 >> 4];
+    if (codepoint_len > len) {
+	return codepoint_status{.more_bytes_needed = uint8_t(codepoint_len - len)};
```
I think there's a bug here. Even if codepoint_len > len, you still have "len" to check. I think you need to check that all the bytes that you do have after the first one are in the range 0x80 and 0xBF, and if not, set "error".
In fact, that's *all* you need to do - it seems to me most of the various cases below are not really needed (more on this below), and moreover, the whole idea of remembering the end of the buffer is not needed. See more on that too.

avikivity replied:
Ah, I think it's protecting against overlong sequences.

nyh replied:
Oh, I *think* maybe I understand now... If you have a 4-byte sequence, and only got 3, you don't check these 3 at all. You just collect them to later, and work on the 4 together, later. Is that true? If so, than the fact the code returns more_bytes_needed immediately, without checking anything, is correct, and there is no bug.

avikivity replied:
Yes, checking is performed when we have a full codepoint.

nyh commented on utils/utf8.cc:
```diff
@@ -99,20 +108,26 @@ static inline std::optional<size_t> validate_naive(const uint8_t *data, size_t l
+        data += cs.bytes_validated;
+        len -= cs.bytes_validated;
+        if (cs.error || cs.more_bytes_needed) {
+            return pos;
```
I didn't understand what you're trying to do here or why.
It seems to me you are trying to return the end of the last full UTF-8 codepoint, perhaps so we can take the last few bytes plus a few bytes from the next buffer to create the last codepoint. But I think that's not needed! The only thing we need to check is that the next "more_bytes_needed" bytes in the next buffer are between 0x80 and 0xBF. That's it! We don't need to remember what the old buffer contained! We just need to remember more_bytes_needed and that's it.

avikivity replied:
No. UTF8 forbids overlong sequences,  0x80-0xbf is too weak.

nyh replied:
I see. I didn't understand this earlier. I am still not sure how such "overlong" sequences can be caused by anything except leading zeros, i.e., 0x80, although I see now https://tools.ietf.org/html/rfc3629#section-10 has another example, and I guess this code was carefully written with that in mind, we we can't try to "improve" or "simplify" it as I have been suggesting to do :-(

nyh reviewed (changes requested):
I think the validator was missing some important checks and didn't understand other checks it did - see my review. If any of the potential bugs I raised are true, it also means the tests missed them so they aren't exhaustive enough.
Moreover, I think the fragmented utf8-checking should actually be simpler than the way you did it. If a UTF-8 "code point" is split between buffers, there is no reason to merge the two halfs. The first half includes the first byte - which indicates the code's **length**, and all the rest of the bytes simply need to begin by bits 10 - i.e., be between 0x80 and 0xBF. In other words, in the second buffer, we just need to check that the first M (calculated in the previous buffer) bytes are between 0x80 and 0xBF - there is no need to keep any byte data from the previous buffer.

haaawk commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
             // C2..DF, 80..BF
-            bytes = 2;
-        } else if (len >= 3) {
+            return codepoint_status{.bytes_validated = codepoint_len};
```
That's my understanding too.

avikivity replied:
Yes, I used the variable so the compiler can consolidate all the return statements.

avikivity commented on utils/utf8.cc:
```diff
@@ -56,22 +56,30 @@ namespace utils {
             // 00..7F
-            bytes = 1;
-        } else if (len >= 2 && byte1 >= 0xC2 && byte1 <= 0xDF &&
+            return codepoint_status{.bytes_validated = codepoint_len};
```
This is only called on buffer edges. The main buffer is evaluated using SSE code.

Note the original author updated the code, and there are now contributed AVX2 implementations. If we want to optimize, we should merge from there.

avikivity commented:
Updated:
 - fixed grammar in changelog
 - added const to len_from_first_nibble array

I did not make any of the simplifications suggested. First, they are wrong, the extra checks are needed to prevent forbidden overlong sequences. Second, the focus of this series is to remove linearization from the validation path, not to optimize the validation code.

avikivity commented:
Updated:
 - made byte1 in evaluate_codepoint() const uint8_t instead of auto

nyh reviewed (approved):
Ok, I withdraw all my requests and hopes for "simplification" of this code, due to https://tools.ietf.org/html/rfc3629#section-10 :-(

denesb commented on utils/utf8.hh:
```diff
@@ -64,6 +65,54 @@ inline std::optional<size_t> validate_with_error_position(bytes_view string) {
+                --len;
+            }
+            if (!partial_more_needed) {
+                // We accumulated a codepoint that straddled two oe more fragments,
```
s/oe/or/

avikivity replied:
Fixed.

avikivity commented:
Updated: fixed a typo in a comment.

avikivity commented:
I dequeued it, some debug mode test fails.

avikivity commented:
Updated:
 - restored patch 1's checks for len. Checking codepoint_len isn't sufficient, because for invalid first bytes have codepoint_len == 1, but the code structure still reads bytes 2..4.
 - rearranged patch 1 to have a smaller diff, so it can more easily be verified
 - replaced a tab that snuck in with spaces
